MADE/made.jl

```diff
@@ -0,0 +1,140 @@
+#=
+Implements Masked AutoEncoder for Density Estimation, by Germain et al. 2015
+Re-implementation by Andrej Karpathy based on https://arxiv.org/abs/1502.03509
+Re-Re-implementation by Tejan Karmali using Flux.jl ;)
+=#
+
+using Flux
+using Flux: glorot_uniform, @epochs
+using NPZ
+
+# ------------------------------------------------------------------------------
+
+add_dims_r(a) = reshape(a, size(a)..., 1)
+add_dims_l(a) = reshape(a, 1, size(a)...)
+
+# ------------------------------------------------------------------------------
+
+struct MaskedDense{F,S,T,M}
+  # same as Linear except has a configurable mask on the weights
+  W::S
+  b::T
+  mask::M
+  σ::F
+end
+
+function MaskedDense(in::Integer, out::Integer; σ = identity)
+  return MaskedDense(param(glorot_uniform(out, in)), param(zeros(out)), ones(out, in), σ)
+end
+
+Flux.treelike(MaskedDense)
+
+function (a::MaskedDense)(x)
+  a.σ.(a.mask .* a.W * x .+ a.b)
+end
+
+function set_mask(a::MaskedDense, mask)
+  a.mask = mask
+end
```

Comment:
Unless there's validation that you need to do this probably doesn't need to be a setter method and could just be replaced at it so call locations by directly setting the attribute?